File: src/main/java/org/athenian/helloworld/HelloWorldClient.java
package org.athenian.helloworld;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;

import java.io.Closeable;
import java.io.IOException;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import static java.lang.String.format;

public class HelloWorldClient
    implements Closeable {
    private final ManagedChannel channel;
    private final GreeterGrpc.GreeterBlockingStub blockingStub;
    private final GreeterGrpc.GreeterStub asyncStub;

    public HelloWorldClient(String host, int port) {
        this(ManagedChannelBuilder.forAddress(host, port)
            // Channels are secure by default (via SSL/TLS). For the example we disable TLS to avoid
            // needing certificates.
            .usePlaintext()
            .build());
    }

    public HelloWorldClient(ManagedChannel channel) {
        this.channel = channel;
        this.blockingStub = GreeterGrpc.newBlockingStub(channel);
        this.asyncStub = GreeterGrpc.newStub(channel);
    }

    public static void main(String... args)
        throws Exception {

        /* Access a service running on the local machine on port 50051 */
        try (HelloWorldClient client = new HelloWorldClient("localhost", 50051)) {
            String name = "world";

            if (args.length > 0)
                name = args[0]; /* Use the arg as the name to greet if provided */

            client.sayHello(name);
            client.sayHelloWithManyRequests(name);
            client.sayHelloWithManyReplies(name);
            client.sayHelloWithManyRequestsAndReplies(name);
        }
    }

    public void sayHello(String name) {
        HelloRequest request = HelloRequest.newBuilder().setName(name).build();
        HelloReply response = this.blockingStub.sayHello(request);
        System.out.printf("sayHello() response: %s\n%n", response.getMessage());
    }

    public void sayHelloWithManyRequests(String name) {

        final CountDownLatch finishLatch = new CountDownLatch(1);

        StreamObserver<HelloReply> responseObserver =
            new StreamObserver<HelloReply>() {
                @Override
                public void onNext(HelloReply reply) {
                    System.out.printf("sayHelloWithManyRequests() response: %s\n%n", reply.getMessage());
                }

                @Override
                public void onError(Throwable t) {
                    Status status = Status.fromThrowable(t);
                    System.out.printf("sayHelloWithMayRequests() failed: %s%n", status);
                    finishLatch.countDown();
                }

                @Override
                public void onCompleted() {
                    finishLatch.countDown();
                }
            };

        StreamObserver<HelloRequest> requestObserver = asyncStub.sayHelloWithManyRequests(responseObserver);

        try {
            for (int i = 0; i < 5; i++) {
                HelloRequest request = HelloRequest.newBuilder().setName(format("%s-%d", name, i)).build();
                requestObserver.onNext(request);

                if (finishLatch.getCount() == 0) {
                    // RPC completed or errored before we finished sending.
                    // Sending further requests won't error, but they will just be thrown away.
                    return;
                }
            }
        } catch (RuntimeException e) {
            // Cancel RPC
            requestObserver.onError(e);
            throw e;
        }

        // Mark the end of requests
        requestObserver.onCompleted();

        // Receiving happens asynchronously
        try {
            finishLatch.await(1, TimeUnit.MINUTES);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public void sayHelloWithManyReplies(String name) {
        HelloRequest request = HelloRequest.newBuilder().setName(name).build();

        System.out.println("sayHelloWithManyReplies() responses:");
        blockingStub.sayHelloWithManyReplies(request).forEachRemaining((reply) -> System.out.println(reply.getMessage()));
        System.out.println();
    }

    public void sayHelloWithManyRequestsAndReplies(String name) {

        final CountDownLatch finishLatch = new CountDownLatch(1);

        StreamObserver<HelloReply> responseObserver =
            new StreamObserver<HelloReply>() {
                @Override
                public void onNext(HelloReply reply) {
                    System.out.printf("sayHelloWithManyRequestsAndReplies() response: %s%n", reply.getMessage());
                }

                @Override
                public void onError(Throwable t) {
                    Status status = Status.fromThrowable(t);
                    System.out.printf("sayHelloWithManyRequestsAndReplies() failed: %s%n", status);
                    finishLatch.countDown();
                }

                @Override
                public void onCompleted() {
                    finishLatch.countDown();
                }
            };

        StreamObserver<HelloRequest> requestObserver = asyncStub.sayHelloWithManyRequestsAndReplies(responseObserver);

        try {
            for (int i = 0; i < 5; i++) {
                HelloRequest request = HelloRequest.newBuilder().setName(format("%s-%d", name, i)).build();
                System.out.printf("sayHelloWithManyRequestsAndReplies() request: %s%n", request.getName());
                requestObserver.onNext(request);

                if (finishLatch.getCount() == 0) {
                    // RPC completed or errored before we finished sending.
                    // Sending further requests won't error, but they will just be thrown away.
                    return;
                }
            }
        } catch (RuntimeException e) {
            // Cancel RPC
            requestObserver.onError(e);
            throw e;
        }

        // Mark the end of requests
        requestObserver.onCompleted();

        // Receiving happens asynchronously
        try {
            finishLatch.await(1, TimeUnit.MINUTES);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void close()
        throws IOException {
        try {
            channel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            throw new IOException(e.getMessage());
        }
    }
}

File: src/main/kotlin/org/athenian/kotlin_helloworld/HelloWorldClient.kt
package org.athenian.kotlin_helloworld

import io.grpc.ConnectivityState
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import io.grpc.Status
import io.grpc.StatusRuntimeException
import io.grpc.stub.StreamObserver
import io.opencensus.contrib.grpc.metrics.RpcViews
import io.opencensus.exporter.stats.prometheus.PrometheusStatsCollector
import io.opencensus.trace.Tracing
import io.prometheus.client.exporter.HTTPServer
import org.athenain.helloworld.GreeterGrpc
import org.athenain.helloworld.HelloReply
import org.athenain.helloworld.HelloRequest
import java.io.Closeable
import java.lang.Thread.sleep
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit


class HelloWorldClient internal constructor(private val channel: ManagedChannel) : Closeable {
    private val blockingStub: GreeterGrpc.GreeterBlockingStub = GreeterGrpc.newBlockingStub(channel)
    private val asyncStub: GreeterGrpc.GreeterStub = GreeterGrpc.newStub(channel)

    constructor(host: String, port: Int = 50051) :
            this(ManagedChannelBuilder.forAddress(host, port)
                         // Channels are secure by default (via SSL/TLS). For the example we disable TLS to avoid
                         // needing certificates.
                         .usePlaintext()
                         .build())

    init {
        channel.notifyWhenStateChanged(ConnectivityState.CONNECTING) { println("Connecting: ${channel.getState(false)}") }
        channel.notifyWhenStateChanged(ConnectivityState.READY) { println("Ready: ${channel.getState(false)}") }
        channel.notifyWhenStateChanged(ConnectivityState.IDLE) { println("Idle: ${channel.getState(false)}") }
    }

    @Throws(InterruptedException::class)
    fun shutdown() {
        channel.shutdown().awaitTermination(5, TimeUnit.SECONDS)
    }

    override fun close() {
        shutdown()
    }

    fun sayHello(name: String) {
        tracer.spanBuilder("grpc-helloworld.sayHello").startScopedSpan()
                .use {
                    try {
                        val request = HelloRequest.newBuilder().setName(name).build()
                        val response = this.blockingStub.sayHello(request)
                        println("sayHello() response: ${response.message}")
                    } catch (e: StatusRuntimeException) {
                        println("sayHello() failed: ${e.status}")
                    }
                }
    }

    fun sayHelloWithManyRequests(name: String) {
        val finishLatch = CountDownLatch(1)
        val responseObserver =
                object : StreamObserver<HelloReply> {
                    override fun onNext(reply: HelloReply) {
                        println("sayHelloWithManyRequests() response: ${reply.message}")
                    }

                    override fun onError(t: Throwable) {
                        val status = Status.fromThrowable(t)
                        println("sayHelloWithMayRequests() failed: $status")
                        finishLatch.countDown()
                    }

                    override fun onCompleted() {
                        finishLatch.countDown()
                    }
                }

        val requestObserver = asyncStub.sayHelloWithManyRequests(responseObserver)

        try {
            repeat(5) {
                val request = HelloRequest.newBuilder()
                        .setName("$name-$it")
                        .build()
                requestObserver.onNext(request)

                if (finishLatch.count == 0L) {
                    // RPC completed or errored before we finished sending.
                    // Sending further requests won't error, but they will just be thrown away.
                    return
                }
            }
        } catch (e: RuntimeException) {
            // Cancel RPC
            requestObserver.onError(e)
            throw e
        }

        // Mark the end of requests
        requestObserver.onCompleted()

        // Receiving happens asynchronously
        try {
            finishLatch.await(1, TimeUnit.MINUTES)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    fun sayHelloWithManyReplies(name: String) {
        val request = HelloRequest.newBuilder().setName(name).build()
        val replies = blockingStub.sayHelloWithManyReplies(request)

        println("sayHelloWithManyReplies() responses:")
        replies.forEach { reply -> println(reply.message) }
        println()
    }

    fun sayHelloWithManyRequestsAndReplies(name: String) {
        val finishLatch = CountDownLatch(1)
        val responseObserver =
                object : StreamObserver<HelloReply> {
                    override fun onNext(reply: HelloReply) {
                        println("sayHelloWithManyRequestsAndReplies() response: ${reply.message}")
                    }

                    override fun onError(t: Throwable) {
                        val status = Status.fromThrowable(t)
                        println("sayHelloWithManyRequestsAndReplies() failed: $status")
                        finishLatch.countDown()
                    }

                    override fun onCompleted() {
                        finishLatch.countDown()
                    }
                }

        val requestObserver = asyncStub.sayHelloWithManyRequestsAndReplies(responseObserver)

        try {
            repeat(5) {
                val request = HelloRequest.newBuilder().setName("$name-$it").build()
                println("sayHelloWithManyRequestsAndReplies() request: ${request.name}")
                requestObserver.onNext(request)

                if (finishLatch.count == 0L) {
                    // RPC completed or errored before we finished sending.
                    // Sending further requests won't error, but they will just be thrown away.
                    return
                }
            }
        } catch (e: RuntimeException) {
            // Cancel RPC
            requestObserver.onError(e)
            throw e
        }

        // Mark the end of requests
        requestObserver.onCompleted()

        // Receiving happens asynchronously
        try {
            finishLatch.await(1, TimeUnit.MINUTES)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    companion object {
        private val tracer = Tracing.getTracer()

        @JvmStatic
        fun main(args: Array<String>) {

            PrometheusStatsCollector.createAndRegister()
            RpcViews.registerClientGrpcViews()
            val http = HTTPServer("localhost", 8889, true)

            /* Access a service running on the local machine on port 50051 */
            /* Use the arg as the name to greet if provided */
            val name = if (args.isNotEmpty()) args[0] else "world"

            HelloWorldClient("localhost")
                    .use { client ->
                        repeat(10) {
                            client.sayHello(name)
                            client.sayHelloWithManyRequests(name)
                            client.sayHelloWithManyReplies(name)
                            client.sayHelloWithManyRequestsAndReplies(name)

                            sleep(1000)
                        }
                    }

            http.stop()
        }
    }
}
